### PR TITLE
Use synchronized map for memory datastore concurrent access [GEOT-2515]

### PR DESCRIPTION
**Pull request by jodygarnett**
See https://osgeo-org.atlassian.net/browse/GEOT-2515

### CODE REVIEW

**Comment by jodygarnett** [comment]
This change is suitable for back port.

File: modules/library/data/src/main/java/org/geotools/data/memory/MemoryEntry.java [hunk 1]
```diff
@@ -70,4 +71,24 @@ protected MemoryState createContentState(ContentEntry entry) {
...
         return "MemoryEntry '" + getTypeName()+"': "+memory.size() + " features";
     }
+    
+    /**
+     * Safely add feature to {@link #memory}.
+     * <p>
+     * Feature is required to be non-null, and of the expected {@link #schema}.
+     * @param feature
+     */
+    void addFeature(SimpleFeature feature) {
```
Comment (tbarsballe) [inline modules/library/data/src/main/java/org/geotools/data/memory/MemoryEntry.java]: Do we want/need an equivalently synchronized removeFeature method?

File: modules/library/data/src/main/java/org/geotools/data/memory/MemoryEntry.java [hunk 2]
```diff
@@ -60,7 +61,7 @@
     MemoryEntry( MemoryDataStore store, SimpleFeatureType schema){
```
Comment (tbarsballe) [inline modules/library/data/src/main/java/org/geotools/data/memory/MemoryEntry.java]: To enforce the synchronized functionality, should the `memory` map be made private?

Comment (jodygarnett) [inline modules/library/data/src/main/java/org/geotools/data/memory/MemoryEntry.java]: tricky question, I would love to say "yes", but I understand that some developers have extended this class (for example to preserve feature ids on transaction commit).

It is currently package visible; so they are unlike to get access to it - so subclasses are unlikely to get direct access to it.

**Comment by jodygarnett** [comment]
merged to master as 6bf42a0396b1b928eb16b9acee27bd108964840c

**Comment by jodygarnett** [comment]
This change is a bug fix, due to the API change of MemoryEntry.getMemory() it will not be back ported.

**Pull request closed without merging**